docs/proposals/0000-support-interchangeable-build-and-run-implementations.md

```diff
@@ -0,0 +1,248 @@
+# Meta
+[meta]: #meta
+- Name: Support Interchangeable Build and Run Implementations
+- Start Date: 2022-04-18
+- Author(s): @matt-royal
+- Status: Draft
+- RFC Pull Request: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This RFC proposes changing the Korifi architecture so that the software that orchestrates the build and run steps is
+outside the `korifi` codebase, so that these flows can be customized and implemented in different ways. To support this,
+we will add the new `BuildWorkload` and `RunWorkload` resources, which contain all information necessary to execute
+their respective processes. The `korifi-controllers` will create these resources, and separate (interchangeable)
+components will reconcile them to build images and run long-running processes. We will provide default implementations
+of these reconcilers that work the same way Korifi currently works (via kpack Images and statefulsets,
+respectively), but operators can choose to deploy alternatives that implement these flows differently.
+
+# Motivation
+[motivation]: #motivation
+
+The build and run implementations are currently baked into korifi-controllers and cannot be changed. While the current
+implementation of these flows may be sufficient for most use cases, there are many reasons operators may want to swap in
+alternate logic. Some examples:
+
+- Using `Deployment`, knative, or other alternative resources for run workloads
+- Signing container images in the build step, and requiring signed images in the run step
+- Security scanning container images during the build and/or run steps
+- Applying conventions to pod spec, such as labels, annotations, volume mounts, etc.
+
+We could accomplish this flexibility with a smaller change: splitting out the current reconcilers for `CFBuild` and
+`CFProcess` into their own repositories, so they can be replaced with alternatives. The issue with this change is that
+the existing reconcilers use information from many resources and an alternative implementation would have to duplicate
+this complex logic. For example, to create the LRPs that currently generate the run step's stateful set, the `CFProcess`
```

Comment:
I think it's worth mentioning that duplicating this logic is not a one time cost either. Any new features we add or bugs we fix in our logic here will need to be carried forward to all implementors. I think the dedicated resource as an interface will make both our own lives and those of anyone extending this easier.